replay/data_preparator.py

```diff
@@ -186,6 +190,209 @@ def _reindex(self, df: DataFrame, entity: str):
+        df.coalesce(1).write.mode("overwrite").json(join(path, "init_args.json"))
+
+        self.instance.user_col_2_index_map.write.mode("overwrite").save(join(path, "user_col_2_index_map.parquet"))
+        self.instance.item_col_2_index_map.write.mode("overwrite").save(join(path, "item_col_2_index_map.parquet"))
+
+
+class JoinIndexerMLReader(MLReader):
+    def load(self, path):
+        """Load the ML instance from the input path."""
+
+        spark = SparkSession.getActiveSession()
+        args = spark.read.json(join(path, "init_args.json")).first().asDict(recursive=True)
+        user_col_2_index_map = spark.read.parquet(join(path, "user_col_2_index_map.parquet"))
+        item_col_2_index_map = spark.read.parquet(join(path, "item_col_2_index_map.parquet"))
+
+        indexer = JoinIndexerTransformer(
+            user_col=args["user_col"],
+            user_type=args["user_type"],
+            user_col_2_index_map=user_col_2_index_map,
+            item_col=args["item_col"],
+            item_type=args["item_type"],
+            item_col_2_index_map=item_col_2_index_map,
+
+        )
+
+        return indexer
+
+
+class JoinIndexerTransformer(Transformer, MLWritable, MLReadable):
+    def __init__(
+            self,
+            user_col: str,
+            item_col: str,
+            user_type: str,
+            item_type: str,
+            user_col_2_index_map: DataFrame,
+            item_col_2_index_map: DataFrame,
+            update_map_on_transform: bool = False
+    ):
+        self.user_col = user_col
+        self.item_col = item_col
+        self.user_type = user_type
+        self.item_type = item_type
+        self.user_col_2_index_map = user_col_2_index_map
+        self.item_col_2_index_map = item_col_2_index_map
+        self.update_map_on_transform = update_map_on_transform
+
+    @property
+    def _init_args(self):
+        return {
+            "user_col": self.user_col,
+            "item_col": self.item_col,
+            "user_type": self.user_type,
+            "item_type": self.item_type,
+            "update_map_on_transform": self.update_map_on_transform
+        }
+
+    def set_update_map_on_transform(self, value: bool):
+        """Sets 'update_map_on_transform' flag"""
+        self.update_map_on_transform = value
+
+    def write(self) -> MLWriter:
+        """Returns MLWriter instance that can save the Transformer instance."""
+        return JoinIndexerMLWriter(self)
+
+    @classmethod
+    def read(cls):
+        """Returns an MLReader instance for this class."""
+        return JoinIndexerMLReader()
+
+    def _update_maps(self, df: DataFrame):
+
+        new_items = (
+            df.join(self.item_col_2_index_map, on=self.item_col, how="left_anti")
+            .select(self.item_col).distinct()
+        )
+        prev_item_count = self.item_col_2_index_map.count()
+        new_items_map = (
+            IndexerEstimator.get_map(new_items, self.item_col, "item_idx")
```

Comment:
get_map might be a part of a common class IndexerHelper inherited by both: IndexerEstimator and JoinIndexerTransformer. This way you still have the classes open for extended and redefining get_map behavior.
But let it be as it is. No need for corrections right now.